dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -91,38 +292,71 @@ void DistributedBlockOutputStream::writeSplit(const Block & block)
             splitted_blocks[shard_idx].getByPosition(col_idx_in_block).column = std::move(splitted_columns[shard_idx]);
     }
 
+    return splitted_blocks;
+}
+
+
+void DistributedBlockOutputStream::writeSplit(const Block & block)
+{
+    Blocks splitted_blocks = splitBlock(block);
+    const size_t num_shards = splitted_blocks.size();
+
     for (size_t shard_idx = 0; shard_idx < num_shards; ++shard_idx)
         if (splitted_blocks[shard_idx].rows())
             writeImpl(splitted_blocks[shard_idx], shard_idx);
+
+    ++blocks_inserted;
 }
 
 
 void DistributedBlockOutputStream::writeImpl(const Block & block, const size_t shard_id)
 {
     const auto & shard_info = cluster->getShardsInfo()[shard_id];
+    size_t finished_writings_count = 0;
     if (shard_info.getLocalNodeCount() > 0)
-        writeToLocal(block, shard_info.getLocalNodeCount());
+        writeToLocal(block, shard_info.getLocalNodeCount(), finished_writings_count);
 
     /// dir_names is empty if shard has only local addresses
     if (!shard_info.dir_names.empty())
         writeToShard(block, shard_info.dir_names);
 }
 
 
-void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_t repeats)
+void DistributedBlockOutputStream::writeToLocal(const Block & block, const size_t repeats, size_t & finished_writings_count)
 {
     InterpreterInsertQuery interp{query_ast, storage.context};
 
     auto block_io = interp.execute();
     block_io.out->writePrefix();
 
     for (size_t i = 0; i < repeats; ++i)
+    {
         block_io.out->write(block);
+        ++finished_writings_count;
+    }
 
     block_io.out->writeSuffix();
 }
 
 
+void DistributedBlockOutputStream::writeToShardSync(
+    const Block & block, const Cluster::ShardInfo & shard_info, size_t replica_id)
+{
+    const auto & dir_name = shard_info.dir_names[replica_id];
```

Comment:
Not very good, that we refer to "directory names" when there are no directories.